### PR TITLE
Add ConnectorId to enforce separation from catalog name

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/CallTask.java [hunk 1]
```diff
@@ -72,6 +74,10 @@ public String getName()
 
         Session session = stateMachine.getSession();
         QualifiedObjectName procedureName = createQualifiedObjectName(session, call, call.getName());
+        ConnectorId connectorId = metadata.getCatalogNames().get(procedureName.getCatalogName());
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/CallTask.java]: Maybe add a `getRequiredCatalog`?

File: presto-main/src/main/java/com/facebook/presto/operator/index/IndexLoader.java [hunk 2]
```diff
@@ -56,6 +57,7 @@
 @ThreadSafe
 public class IndexLoader
 {
+    private static final ConnectorId CONNECTOR_ID = new ConnectorId("connector_id");
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/index/IndexLoader.java]: Should this be `$index`?

File: presto-main/src/main/java/com/facebook/presto/server/PrestoServer.java [hunk 3]
```diff
@@ -156,15 +157,17 @@ private static void updateConnectorIds(Announcer announcer, Metadata metadata, S
 
         // automatically build connectorIds if not configured
         if (connectorIds.isEmpty()) {
-            Set<String> catalogs = metadata.getCatalogNames().keySet();
+            Map<String, ConnectorId> catalogNames = metadata.getCatalogNames();
             // if this is a dedicated coordinator, only add jmx
             if (serverConfig.isCoordinator() && !schedulerConfig.isIncludeCoordinator()) {
-                if (catalogs.contains("jmx")) {
-                    connectorIds.add("jmx");
+                if (catalogNames.containsKey("jmx")) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoServer.java]: I think we can remove these lines

**Comment by electrum** [comment]
Can we remove `TableMetadata`?

File: presto-main/src/test/java/com/facebook/presto/execution/BenchmarkNodeScheduler.java [hunk 4]
```diff
@@ -87,6 +88,7 @@
     private static final int RACKS = DATA_NODES / 25;
     private static final int SPLITS = NODES * (MAX_SPLITS_PER_NODE + MAX_PENDING_SPLITS_PER_TASK_PER_NODE / 3);
     private static final int SPLIT_BATCH_SIZE = 100;
+    private static final ConnectorId CONNECTOR_ID = new ConnectorId("connector_id");
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/execution/BenchmarkNodeScheduler.java]: `test`?

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 5]
```diff
@@ -248,15 +250,17 @@ public void createCatalog(String catalogName, String connectorName)
...
-        log.info("Created catalog %s in %s", catalogName, nanosSince(start).convertToMostSuccinctTimeUnit());
+        ConnectorId connectorId = Iterables.getOnlyElement(connectorIds);
+        log.info("Created catalog %s (%s) in %s", catalogName, connectorId, nanosSince(start).convertToMostSuccinctTimeUnit());
 
         // wait for all nodes to announce the new catalog
         start = System.nanoTime();
-        while (!isConnectionVisibleToAllNodes(catalogName)) {
-            Assertions.assertLessThan(nanosSince(start), new Duration(100, SECONDS), "waiting form connector " + connectorName + " to be initialized in every node");
+        while (!isConnectionVisibleToAllNodes(connectorId)) {
+            Assertions.assertLessThan(nanosSince(start), new Duration(100, SECONDS), "waiting form connector " + connectorId + " to be initialized in every node");
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: "waiting for"

Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: Maybe use `format`

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 6]
```diff
@@ -248,15 +250,17 @@ public void createCatalog(String catalogName, String connectorName)
     public void createCatalog(String catalogName, String connectorName, Map<String, String> properties)
     {
         long start = System.nanoTime();
+        Set<ConnectorId> connectorIds = new HashSet<>();
         for (TestingPrestoServer server : servers) {
-            server.createCatalog(catalogName, connectorName, properties);
+            connectorIds.add(server.createCatalog(catalogName, connectorName, properties));
         }
-        log.info("Created catalog %s in %s", catalogName, nanosSince(start).convertToMostSuccinctTimeUnit());
+        ConnectorId connectorId = Iterables.getOnlyElement(connectorIds);
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: Static import if not already used

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 7]
```diff
@@ -248,15 +250,17 @@ public void createCatalog(String catalogName, String connectorName)
...
+        log.info("Created catalog %s (%s) in %s", catalogName, connectorId, nanosSince(start).convertToMostSuccinctTimeUnit());
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: `. convertToMostSuccinctTimeUnit()` no longer needed

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 8]
```diff
@@ -354,10 +355,10 @@ public ClientSession toClientSession(URI server, boolean debug, Duration clientR
     {
         ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
         properties.putAll(systemProperties);
-        for (Entry<String, Map<String, String>> catalogProperties : this.catalogProperties.entrySet()) {
-            String catalog = catalogProperties.getKey();
-            for (Entry<String, String> entry : catalogProperties.getValue().entrySet()) {
-                properties.put(catalog + "." + entry.getKey(), entry.getValue());
+        for (Entry<ConnectorId, Map<String, String>> connectorProperties : this.connectorProperties.entrySet()) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: `connectorProperties` might not needed `this`

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: Rename `connectorProperties` to `entry` and assign key and value to variables

File: presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java [hunk 9]
```diff
@@ -87,68 +90,101 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
...
-    private <T> void addSessionPropertyInternal(Optional<String> catalogName, PropertyMetadata<T> sessionProperty)
+    public Optional<ConnectorId> getSessionPropertyConnectorId(String name)
     {
-        SessionProperty<T> value = new SessionProperty<>(catalogName, sessionProperty);
-        allSessionProperties.put(value.getFullyQualifiedName(), value);
+        requireNonNull(name, "name is null");
+
+        SessionPropertyMetadata<?> sessionPropertyMetadata = allSessionProperties.get(name);
+        if (sessionPropertyMetadata == null) {
+            throw new PrestoException(INVALID_SESSION_PROPERTY, "Unknown session property " + name);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java]: ` "Unknown session property :"`

**Comment by electrum** [comment]
First three commits look good

File: presto-main/src/main/java/com/facebook/presto/connector/ConnectorId.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ConnectorId
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/connector/ConnectorId.java]: Make this final

File: presto-main/src/main/java/com/facebook/presto/event/query/QueryMonitor.java [hunk 11]
```diff
@@ -136,7 +136,7 @@ public void queryCompletedEvent(QueryInfo queryInfo)
             ImmutableList.Builder<QueryInputMetadata> inputs = ImmutableList.builder();
             for (Input input : queryInfo.getInputs()) {
                 inputs.add(new QueryInputMetadata(
-                        input.getConnectorId(),
+                        input.getConnectorId().toString(),
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/event/query/QueryMonitor.java]: Should these be `getCatalogName()`? The methods do the same thing, but it's not clear why the usages are mixed.

File: presto-main/src/main/java/com/facebook/presto/operator/index/IndexLoader.java (hunk 2 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/index/IndexLoader.java]: Maybe call this `INDEX_CONNECTOR_ID`

File: presto-main/src/main/java/com/facebook/presto/server/PrestoServer.java (hunk 3 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoServer.java]: If the JMX part is removed, the `if` can be inverted

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java [hunk 12]
```diff
@@ -408,7 +410,7 @@ private static void updateConnectorIdAnnouncement(Announcer announcer, String co
         Map<String, String> properties = new LinkedHashMap<>(announcement.getProperties());
         String property = nullToEmpty(properties.get("connectorIds"));
         Set<String> connectorIds = new LinkedHashSet<>(Splitter.on(',').trimResults().omitEmptyStrings().splitToList(property));
-        connectorIds.add(connectorId);
+        connectorIds.add(connectorId.toString());
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: Use `getCatalogName()`?

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 13]
```diff
@@ -1177,9 +1183,10 @@ private void assertFailsWithoutExperimentalSyntax(SemanticErrorCode error, @Lang
         }
     }
 
-    private static Connector createTestingConnector(String connectorId)
+    @SuppressWarnings("deprecation")
+    private static Connector createTestingConnector(ConnectorId connectorId)
     {
-        return new LegacyTransactionConnector(connectorId, new com.facebook.presto.spi.Connector()
+        return new LegacyTransactionConnector(connectorId.toString(), new com.facebook.presto.spi.Connector()
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java]: `getCatalogName()`?

File: presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestRaptorMetadata.java [hunk 14]
```diff
@@ -109,7 +110,7 @@ public void setupDatabase()
 
         RaptorConnectorId connectorId = new RaptorConnectorId("raptor");
         InMemoryNodeManager nodeManager = new InMemoryNodeManager();
-        nodeManager.addCurrentNodeConnector(connectorId.toString());
+        nodeManager.addCurrentNodeConnector(new ConnectorId(connectorId.toString()));
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestRaptorMetadata.java]: `getCatalogName()`?

File: presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestRaptorSplitManager.java [hunk 15]
```diff
@@ -176,7 +177,7 @@ public void testAssignRandomNodeWhenBackupAvailable()
         RaptorConnectorId connectorId = new RaptorConnectorId("raptor");
         NodeSupplier nodeSupplier = new RaptorNodeSupplier(nodeManager, connectorId);
         PrestoNode node = new PrestoNode(UUID.randomUUID().toString(), new URI("http://127.0.0.1/"), NodeVersion.UNKNOWN);
-        nodeManager.addNode(connectorId.toString(), node);
+        nodeManager.addNode(new ConnectorId(connectorId.toString()), node);
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestRaptorSplitManager.java]: `getCatalogName()`?

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 16]
```diff
@@ -265,14 +269,14 @@ public void createCatalog(String catalogName, String connectorName, Map<String,
...
+        log.info("Announced catalog %s (%s) in %s", catalogName, connectorId, nanosSince(start).convertToMostSuccinctTimeUnit());
     }
 
-    private boolean isConnectionVisibleToAllNodes(String connectorId)
+    private boolean isConnectionVisibleToAllNodes(ConnectorId connectorId)
     {
         for (TestingPrestoServer server : servers) {
             server.refreshNodes();
-            Set<Node> activeNodesWithConnector = server.getActiveNodesWithConnector(connectorId);
+            Set<Node> activeNodesWithConnector = server.getActiveNodesWithConnector(connectorId.toString());
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: `getCatalogName()`?

**Comment by electrum** [comment]
Fourth commit looks good. I flagged all the places using `ConnectorId.toString()` since it's not clear if it should be that or `getCatalogName()`.

File: presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java [hunk 17]
```diff
@@ -87,68 +90,101 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
...
-        for (PropertyMetadata<?> sessionProperty : sessionProperties) {
-            addSessionPropertyInternal(Optional.of(catalog), sessionProperty);
+        List<SessionPropertyMetadata<?>> properties = sessionProperties.stream()
+                .map(sessionProperty -> new SessionPropertyMetadata<>(Optional.of(catalog), Optional.of(connectorId), sessionProperty))
+                .collect(toList());
+
+        Map<String, SessionPropertyMetadata<?>> propertiesByName = Maps.uniqueIndex(properties, SessionPropertyMetadata::getPropertyName);
+        checkState(connectorSessionProperties.putIfAbsent(connectorId, propertiesByName) == null, "SessionProperties for connectorId '%s' are already registered", connectorId);
+
+        for (SessionPropertyMetadata<?> property : properties) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java]: Maybe call this `value` so it matches the `.put()` in the other method

File: presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java [hunk 18]
```diff
@@ -87,68 +90,101 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
...
+            allSessionProperties.put(property.getFullyQualifiedName(), property);
         }
     }
 
-    private <T> void addSessionPropertyInternal(Optional<String> catalogName, PropertyMetadata<T> sessionProperty)
+    public Optional<ConnectorId> getSessionPropertyConnectorId(String name)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java]: This method can be
```java
return getSessionPropertyMetadata(name).getConnectorId();

File: presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java [hunk 19]
```diff
@@ -87,68 +90,101 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
...
-        return sessionProperty.getMetadata();
+        return sessionPropertyMetadata;
+    }
+
+    public SessionPropertyMetadata<?> getSessionPropertyMetadata(ConnectorId connectorId, String propertyName)
+    {
+        requireNonNull(connectorId, "connectorId is null");
+        requireNonNull(propertyName, "propertyName is null");
+        Map<String, SessionPropertyMetadata<?>> properties = connectorSessionProperties.get(connectorId);
+        if (properties == null || properties.isEmpty()) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java]: Is this `isEmpty()` check correct? The `addConnectorSessionProperties()` method doesn't validate that the list of properties is non-empty.

File: presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java [hunk 20]
```diff
@@ -87,68 +90,101 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
...
+            throw new PrestoException(INVALID_SESSION_PROPERTY, "Unknown connector " + connectorId);
+        }
+
+        SessionPropertyMetadata<?> propertyMetadata = properties.get(propertyName);
+        if (propertyMetadata == null) {
+            String fullPropertyName = properties.values().iterator().next().getCatalogName()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java]: I see this requires the map to be non-empty

File: presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java [hunk 21]
```diff
@@ -117,33 +115,41 @@ public static Session createSessionForRequest(HttpServletRequest servletRequest,
...
 
         // parse session properties
-        Multimap<String, Entry<String, String>> sessionPropertiesByCatalog = HashMultimap.create();
+        Map<String, String> sessionProperties = new HashMap<>();
         for (String sessionHeader : splitSessionHeader(servletRequest.getHeaders(PRESTO_SESSION))) {
-            parseSessionHeader(sessionHeader, sessionPropertiesByCatalog, sessionPropertyManager);
+            parseSessionHeader(sessionHeader, sessionProperties, sessionPropertyManager);
         }
-
-        // verify user can set the session properties
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java]: Should this stay?

File: presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java [hunk 22]
```diff
@@ -117,33 +115,41 @@ public static Session createSessionForRequest(HttpServletRequest servletRequest,
...
+                else if (nameParts.size() == 2) {
+                    String catalogName = nameParts.get(0);
+                    String propertyName = nameParts.get(1);
+
+                    assertRequest(!catalogName.isEmpty(), "Invalid %s header", PRESTO_SESSION);
+                    assertRequest(!propertyName.isEmpty(), "Invalid %s header", PRESTO_SESSION);
                     accessControl.checkCanSetCatalogSessionProperty(identity, catalogName, propertyName);
+
+                    Optional<ConnectorId> connectorId = sessionPropertyManager.getSessionPropertyConnectorId(entry.getKey());
+                    verify(connectorId.isPresent());
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java]: Why would this occur? What if the property doesn't exist? Should this be `badRequest()`?

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java]: Any reason not to do this check before the access check?

**Comment by electrum** [comment]
Looks good

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java (hunk 12 above)
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: This is the code for announcements and we announce connectorIds (not catalogs)

**Pull request closed without merging**